### PR TITLE
Category task

### PR DESCRIPTION
**Pull request by DmitryLazyuka**

### CODE REVIEW

File: src/main/java/org/example/springproject/controller/BookController.java [hunk 1]
```diff
@@ -52,6 +54,7 @@ public BookDto getBookById(@PathVariable @Positive Long id) {
     @PreAuthorize("hasRole('ADMIN')")
     @PostMapping()
     public BookDto createBook(@RequestBody @Valid CreateBookRequestDto bookDto) {
+
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springproject/controller/BookController.java]:
```suggestion
```

File: src/main/java/org/example/springproject/model/Book.java [hunk 2]
```diff
@@ -34,4 +42,12 @@ public class Book {
     private String coverImage;
     @Column(nullable = false)
     private boolean isDeleted = false;
+    @ManyToMany(fetch = FetchType.LAZY)
+    @JoinTable(
+            name = "book_category",
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springproject/model/Book.java]:
```suggestion
            name = "books_categories",
```

File: src/main/java/org/example/springproject/model/Book.java [hunk 3]
```diff
@@ -34,4 +42,12 @@ public class Book {
     private String coverImage;
     @Column(nullable = false)
     private boolean isDeleted = false;
+    @ManyToMany(fetch = FetchType.LAZY)
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springproject/model/Book.java]:
```suggestion
    @ManyToMany
```

File: src/main/java/org/example/springproject/model/Category.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false)
+    private String name;
+    private String description;
+    @Column(nullable = false)
+    private boolean isDeleted = false;
+
+    public Category() {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springproject/model/Category.java]: use Lombok annotation

File: src/main/java/org/example/springproject/repository/category/CategoryRepository.java [hunk 5]
```diff
@@ -0,0 +1,13 @@
+package org.example.springproject.repository.category;
+
+import java.util.Set;
+import org.example.springproject.model.Category;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
+
+public interface CategoryRepository extends JpaRepository<Category, Long> {
+
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springproject/repository/category/CategoryRepository.java]:
```suggestion
```

File: src/main/java/org/example/springproject/service/CategoryService.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
+package org.example.springproject.service;
+
+import java.util.List;
+import org.example.springproject.dto.category.CategoryDto;
+import org.example.springproject.dto.category.CreateCategoryRequestDto;
+import org.example.springproject.model.Category;
+import org.springframework.data.domain.Pageable;
+
+public interface CategoryService {
+
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springproject/service/CategoryService.java]:
```suggestion
```

File: src/main/java/org/example/springproject/service/impl/BookServiceImpl.java [hunk 7]
```diff
@@ -21,10 +26,14 @@
 public class BookServiceImpl implements BookService {
     private final BookMapper bookMapper;
     private final BookRepository bookRepository;
+    private final CategoryRepository categoryRepository;
     private final BookSpecificationBuilder bookSpecificationBuilder;
 
     @Override
     public BookDto save(CreateBookRequestDto requestDto) {
+        if (!hasValidCategories(requestDto)) {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springproject/service/impl/BookServiceImpl.java]:
```suggestion
        Book book = bookMapper.toBook(bookDto);
        book.setCategories(categoriesIdToCategories(bookDto.categoriesId()));
        return bookMapper.toDto(bookRepository.save(book));
```

where

```
private Set<Category> categoriesIdToCategories(List<Long> categories) {
        return categories.stream()
                .map(categoryRepository::getReferenceById)
                .collect(Collectors.toSet());
    }
```

File: src/main/java/org/example/springproject/service/impl/CategoryServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+@RequiredArgsConstructor
+public class CategoryServiceImpl implements CategoryService {
+    private final CategoryMapper categoryMapper;
+    private final CategoryRepository categoryRepository;
+
+    @Override
+    public List<CategoryDto> findAll(Pageable pageable) {
+        return categoryRepository.findAll(pageable).stream()
+                .map(categoryMapper::toDto)
+                .collect(Collectors.toList());
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springproject/service/impl/CategoryServiceImpl.java]:
```suggestion
                .toList();
```

File: src/main/java/org/example/springproject/service/impl/CategoryServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    @Override
+    public CategoryDto getById(Long id) {
+        Category category = categoryRepository.findById(id)
+                .orElseThrow(() -> new NoSuchElementException(
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springproject/service/impl/CategoryServiceImpl.java]:
```suggestion
                .orElseThrow(() -> new EntityNotFoundException(
```

File: src/main/java/org/example/springproject/service/impl/CategoryServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+                        "There is no category with id " + id));
+        return categoryMapper.toDto(category);
+    }
+
+    @Override
+    public Category getCategoryById(Long id) {
+        return categoryRepository.findById(id)
+                .orElseThrow(() -> new NoSuchElementException(
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springproject/service/impl/CategoryServiceImpl.java]:
```suggestion
                .orElseThrow(() -> new EntityNotFoundException(
```

File: src/main/java/org/example/springproject/service/impl/CategoryServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    @Override
+    public CategoryDto getById(Long id) {
+        Category category = categoryRepository.findById(id)
+                .orElseThrow(() -> new NoSuchElementException(
+                        "There is no category with id " + id));
+        return categoryMapper.toDto(category);
+    }
+
+    @Override
+    public Category getCategoryById(Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springproject/service/impl/CategoryServiceImpl.java]: what the difference between this method and getById ?

File: src/main/java/org/example/springproject/service/impl/CategoryServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+        Optional<Category> foundCategory = categoryRepository.findById(id);
+
+        if (foundCategory.isEmpty()) {
+            throw new NoSuchElementException("There is no category with id " + id);
+        }
+        Category category = foundCategory.get();
+        category.setName(categoryDto.getName());
+        category.setDescription(categoryDto.getDescription());
+        Category updatedCategory = categoryRepository.save(category);
+        return categoryMapper.toDto(updatedCategory);
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springproject/service/impl/CategoryServiceImpl.java]:
```suggestion
        Category category = repository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException(
                        "No such a category with id: " + id
                ));
        mapper.updateCategoryFromDto(categoryDto, category);
        return mapper.toDto(repository.save(category));
```

File: src/main/java/org/example/springproject/repository/category/CategoryRepository.java [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+    @Query("SELECT COUNT(c) FROM Category c WHERE c.id IN :ids")
+    long countByIds(@Param("ids") Set<Long> ids);
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springproject/repository/category/CategoryRepository.java]: remove

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/org/example/springproject/controller/BookController.java [hunk 14]
```diff
@@ -33,6 +34,7 @@
 @RequestMapping("/books")
 public class BookController {
     private final BookService bookService;
+    private final CategoryRepository categoryRepository;
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springproject/controller/BookController.java]: unused

File: src/main/java/org/example/springproject/exception/CustomGlobalExceptionHandler.java [hunk 15]
```diff
@@ -54,4 +55,14 @@ public ResponseEntity<Object> handleRegistrationException(RegistrationException
         body.put("error", ex.getMessage());
         return new ResponseEntity<>(body, HttpStatus.NOT_FOUND);
     }
+
+    @ExceptionHandler(SQLIntegrityConstraintViolationException.class)
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springproject/exception/CustomGlobalExceptionHandler.java]: unused exception

File: src/main/java/org/example/springproject/dto/book/CreateBookRequestDto.java [hunk 16]
```diff
@@ -21,4 +22,5 @@ public class CreateBookRequestDto {
     private BigDecimal price;
     private String description;
     private String coverImage;
+    private Set<Long> categoryIds;
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springproject/dto/book/CreateBookRequestDto.java:27]:
```suggestion
    @NotEmpty
    private Set<Long> categoryIds;
```

File: src/main/java/org/example/springproject/mapper/BookMapper.java [hunk 17]
```diff
@@ -1,14 +1,43 @@
...
+                    .map(Category::getId)
+                    .collect(Collectors.toSet());
+            bookDto.setCategoryIds(categories);
+        }
+    }
+
+    @AfterMapping
+    default void mapCategoryIdsToCategories(@MappingTarget Book book,
+                                            CreateBookRequestDto requestDto) {
+        if (requestDto.getCategoryIds() != null) {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springproject/mapper/BookMapper.java]:
```suggestion
```

File: src/main/java/org/example/springproject/mapper/CategoryMapper.java [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+import org.example.springproject.dto.category.CategoryDto;
+import org.example.springproject.dto.category.CreateCategoryRequestDto;
+import org.example.springproject.model.Category;
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+import org.mapstruct.MappingTarget;
+
+@Mapper(config = MapperConfig.class)
+public interface CategoryMapper {
+
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springproject/mapper/CategoryMapper.java]:
```suggestion
```

File: src/main/java/org/example/springproject/mapper/UserMapper.java [hunk 19]
```diff
@@ -8,6 +8,7 @@
 
 @Mapper(config = MapperConfig.class)
 public interface UserMapper {
+
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springproject/mapper/UserMapper.java]:
```suggestion
```

File: src/main/java/org/example/springproject/service/CategoryService.java (hunk 6 above)
Comment (Elena-Bruyako) [reply src/main/java/org/example/springproject/service/CategoryService.java]: not fixed
 don't begin class or method with empty line

File: src/main/java/org/example/springproject/service/impl/BookServiceImpl.java [hunk 20]
```diff
@@ -50,13 +57,18 @@ public BookDto update(Long id, CreateBookRequestDto requestDto) {
         if (optionalBook.isEmpty()) {
             throw new EntityNotFoundException("Book not found with id: " + id);
         }
+
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springproject/service/impl/BookServiceImpl.java]:
```suggestion
        Book book = bookRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException(getNotFoundMessage(id)));
        bookMapper.updateBookFromDto(bookDto, book);
        book.setCategories(categoriesIdToCategories(bookDto.categoriesId()));
        return bookMapper.toDto(bookRepository.save(book));
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/org/example/springproject/controller/CategoryController.java [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springproject/controller/CategoryController.java:30]:
```suggestion
@RestController
@Validated
```

as you use `@Positive` for Long id

File: src/main/java/org/example/springproject/service/impl/BookServiceImpl.java [hunk 22]
```diff
@@ -75,4 +71,17 @@ public List<BookDto> search(BookSearchParametersDto bookSearchParametersDto,
         return bookRepository.findAll(bookSpecification, pageable).stream()
                 .map(bookMapper::toBookDto).toList();
     }
+
+    @Override
+    public List<BookDtoWithoutCategoryIds> findByCategoryId(Long categoryId) {
+        return bookRepository.findAllByCategories_Id(categoryId).stream()
+                .map(bookMapper::toDtoWithoutCategories)
+                .collect(Collectors.toList());
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/springproject/service/impl/BookServiceImpl.java]:
```suggestion
                .toList();
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by sarakhmen — APPROVED** [review]

**Pull request merged**